Warn if an annotation appears before modifiers such as "public"

mernst opened the pull request:
This has a problem: it runs for every BaseTypeVisitor, so in a type system such as Signedness (which also uses the Value Checker), the warnings are issued twice.  How should functionality be added to the framework that should be executed only once?

wmdietl commented on framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:
```diff
@@ -890,6 +931,23 @@ public Void visitTypeParameter(TypeParameterTree node, Void p) {
 
     @Override
     public Void visitVariable(VariableTree node, Void p) {
```
You probably want the same check on method declarations for return types.

wmdietl commented on framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:
```diff
@@ -890,6 +931,23 @@ public Void visitTypeParameter(TypeParameterTree node, Void p) {
+        List<? extends AnnotationTree> annotations = modifiersTree.getAnnotations();
+        if (!modifierSet.isEmpty() && !annotations.isEmpty()) {
+            AnnotationTree anno = annotations.get(0);
+            if (((JCTree) anno).getStartPosition() == ((JCTree) node).getStartPosition()
```
Would something like this be a more general (yet still hacky) check:
- convert modifierSet into a String, so we get "public static final" or some such.
- keep a counter `nextPos` initialized to that string's length.
- go through each annotation:
  - if it's a declaration annotation: add the length of the annotation to nextPos
    - length of toString of the annotation isn't useful, as it might be different length than original input
    - maybe using JCTree.getEndPosition(EndPosTable) and com.sun.tools.javac.tree.EndPosTable, note requires -Xjcov
  - if it's a type annotation: make sure start pos of annotation > nextPos.
  - once you see a type annotation, don't allow declaration annotations.

Ensuring that no declaration annotation happens after a type annotation should be an easy additional check.
As precise calculations of lengths would require -Xjcov, I'm not sure we would want to do that in the BaseTypeVisitor, but rather write a separate optional checker for it.

mernst commented on framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:
```diff
@@ -209,6 +221,35 @@ protected BaseTypeVisitor(BaseTypeChecker checker, Factory typeFactory) {
+     * @param checker the type-checker associated with this visitor (for callbacks to {@link
+     *     TypeHierarchy#isSubtype})
+     */
+    private BaseTypeVisitor(BaseTypeChecker checker, boolean privateConstructor) {
```
Agreed.  That was an experiment that got accidentally committed to this branch.  Sorry about that.

mernst commented on framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:
```diff
@@ -914,6 +972,31 @@ public Void visitVariable(VariableTree node, Void p) {
+     * Return true if the given annotation is a type annotation: that is, its definition is
+     * meta-annotated with {@code @Target({TYPE_USE,....})}.
+     */
+    private boolean isTypeAnnotation(AnnotationTree anno) {
```
Thanks for pointing out that method.  That method is undocumented, has a weird interface (requires a Class<?>, but just for diagnostic messages), and crashes the Checker Framework if the user ever writes an annotation that is both a type annotation and a declaration annotation.  I think it would need to be cleaned up to be usable for clients.

mernst commented:
@wmdietl I have addressed all your feedback, used the algorithm you suggested, and added more tests.
There is still the concern that each warning is issued twice, in a type system like signedness (which also runs the Constant Value Checker).

wmdietl commented:
#2314 fixes up compiler message comparisons, which should fix your problem with the same message appearing multiple times.

https://github.com/mernst/checker-framework/pull/2 makes a few changes to this PR. Take a look.

mernst commented:
@wmdietl both of the pull requests you referenced (#2314 and mernst#2) fail tests.  Can you fix the tests so they can be reviewed and merged?
